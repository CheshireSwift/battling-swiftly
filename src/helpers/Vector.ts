export class Vector {
  x: number
  y: number

  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  difference(other: Vector): Vector {
    return new Vector(other.x - this.x, other.y - this.y)
  }

  length(): number {
    return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2))
  }

  suffix(suffix: string) {
    return {
      ['x' + suffix]: this.x,
      ['y' + suffix]: this.y,
    }
  }
}
