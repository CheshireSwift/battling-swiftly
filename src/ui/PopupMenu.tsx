import * as React from 'react'
import * as _ from 'lodash'
import Vector from '../helpers/Vector'

type PopupMenuProps<T> = {
  position: { top: number; left: number } | null
  menuItems: Array<[T, React.ReactNode] | null | undefined | false | '' | 0>
  selectedItem?: T
  onSelectItem?: (key: T) => void
}

export const PopupMenu = <T extends any>(props: PopupMenuProps<T>) =>
  props.position ? (
    <div
      style={{
        position: 'absolute',
        pointerEvents: 'all',
        ...props.position,
      }}
    >
      {_.compact(props.menuItems).map(([key, value]) => (
        <div
          key={key.toString()}
          id={'popup-menu-item-' + key}
          style={{
            border: props.selectedItem === key ? '1px solid lime' : 'none',
            background: 'gray',
          }}
          onClick={e => {
            props.onSelectItem && props.onSelectItem(key)
            e.stopPropagation()
          }}
        >
          {value}
        </div>
      ))}
    </div>
  ) : null

export default PopupMenu
