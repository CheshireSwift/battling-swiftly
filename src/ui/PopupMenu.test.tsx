import * as React from 'react'
import { mount, render } from 'enzyme'

import PopupMenu from './PopupMenu'
import Vector from '../helpers/Vector'

describe('the popup menu', () => {
  it('matches the snapshot', () => {
    expect(
      <PopupMenu
        position={{ top: 10, left: 20 }}
        menuItems={[['a', 'foo'], ['b', 'bar']]}
      />,
    ).toMatchSnapshot()
  })

  it('hides itself when not given a position', () => {
    const popupMenu = render(<PopupMenu position={null} menuItems={[]} />)
    expect(popupMenu.html()).toBeNull()
  })

  it('displays the provided items', () => {
    const popupMenu = render(
      <PopupMenu
        position={{ top: 0, left: 0 }}
        menuItems={[['a', 'foo'], ['b', 'bar']]}
      />,
    )
    expect(popupMenu.text()).toContain('foo')
    expect(popupMenu.text()).toContain('bar')
  })

  it('highlights the selected item', () => {
    const popupMenu = render(
      <PopupMenu
        position={{ top: 0, left: 0 }}
        menuItems={[['a', 'foo'], ['b', 'bar'], ['c', 'baz']]}
        selectedItem="a"
      />,
    )

    const aElem = popupMenu.find('*:contains("foo")')
    const bElem = popupMenu.find('*:contains("bar")')
    const cElem = popupMenu.find('*:contains("baz")')

    expect(bElem.attr('style')).not.toEqual(aElem.attr('style'))
    expect(bElem.attr('style')).toEqual(cElem.attr('style'))
  })

  it('fires the selection handler when an item is clicked', () => {
    const onSelectItem = jest.fn()
    const popupMenu = mount(
      <PopupMenu
        position={{ top: 0, left: 0 }}
        menuItems={[['a', 'foo'], ['b', 'bar']]}
        onSelectItem={onSelectItem}
      />,
    )

    popupMenu.findWhere(e => !!e.type() && e.text() === 'foo').simulate('click')

    expect(onSelectItem).toHaveBeenCalledWith('a')
  })

  it('draws itself at the specified position', () => {
    const popupMenu = render(
      <PopupMenu position={new Vector(123, 456).toStyle()} menuItems={[]} />,
    )

    const style = popupMenu.attr('style')
    expect(style).toContain('left:123')
    expect(style).toContain('top:456')
  })

  it('ignores falsy items', () => {
    const popupMenu = render(
      <PopupMenu
        position={new Vector(0, 0).toStyle()}
        menuItems={[[1, 'a'], null, [2, 'b'], '', false, [3, 'c']]}
      />,
    )

    expect(popupMenu.children().length).toBe(3)
  })
})
