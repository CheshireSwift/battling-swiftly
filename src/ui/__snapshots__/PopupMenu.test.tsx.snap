// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the popup menu matches the snapshot 1`] = `
<Unknown
  menuItems={
    Array [
      Array [
        "a",
        "foo",
      ],
      Array [
        "b",
        "bar",
      ],
    ]
  }
  position={
    Object {
      "left": 20,
      "top": 10,
    }
  }
/>
`;
